scripts/shufbel/shufbel.go

```diff
@@ -0,0 +1,128 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"math/rand"
+	"os"
+	"runtime"
+	"sync"
+	"syscall"
+	"unsafe"
+
+	"github.com/vmunoz82/shuffle"
+)
+
+/* Function for each round, could be anything don't need to be reversible */
+func roundFunction(v, key shuffle.FeistelWord) shuffle.FeistelWord {
+	return (v * 941083987) ^ (key >> (v & 7) * 104729)
+}
+
+// main function
+func main() {
+	// parse CLI options
+	threads := flag.Int("threads", runtime.NumCPU(), "number of threads")
+	rseed := flag.Int64("rseed", 0, "random seed")
+	flag.Parse()
+
+	// limit go runtime threads
+	runtime.GOMAXPROCS(*threads)
+	pageSize := syscall.Getpagesize()
+
+	files := flag.Args()
+	if len(files) != 2 {
+		fmt.Println("Usage: shufbel [options] input output ...")
+		flag.PrintDefaults()
+		return
+	}
+	input_file := files[0]
+	output_file := files[1]
+
+	// mmap input file
+	fin, err := os.Open(input_file)
+	if err != nil {
+		fmt.Printf("Error opening file %s: %v\n", input_file, err)
+		return
+	}
+	defer fin.Close()
+
+	fout, err := os.OpenFile(output_file, os.O_RDWR|os.O_CREATE, 0644)
+	if err != nil {
+		fmt.Printf("Error opening file %s: %v\n", output_file, err)
+		return
+	}
+
+	fi, err := fin.Stat()
+	if err != nil {
+		fmt.Printf("Error getting file info: %v\n", err)
+	}
+	sizeBytes := uint64(fi.Size())
+	if sizeBytes == 0 {
+		fmt.Printf("Error: file size is 0: %s\n", input_file)
+		return
+	}
+	if sizeBytes%16 != 0 {
+		fmt.Printf("Error: file size is not divisible by 16")
+		return
+	}
+
+	fout.Truncate(int64(sizeBytes))
+
+	data, err := syscall.Mmap(int(fin.Fd()), 0, int(sizeBytes), syscall.PROT_READ, syscall.MAP_PRIVATE|syscall.MAP_NORESERVE)
+	if err != nil {
+		fmt.Printf("Error mmaping file %s: %v\n", input_file, err)
+		return
+	}
+	if uintptr(unsafe.Pointer(&data[0]))%uintptr(pageSize) != 0 {
+		fmt.Println("mmap returned a non-page-aligned address: %p", uintptr(unsafe.Pointer(&data[0])))
```

Comment:
why aren't you dispatching goroutines to do striped reads of the file

Reply:
I am, scroll down

Reply:
No I specifically mean like a read command, this is supposed to work on graphs that are TB in size. How will this even be mmaped?